File: python/semantic_kernel/processes/kernel_process/kernel_process.py
# Copyright (c) Microsoft. All rights reserved.

from collections.abc import Callable
from typing import TYPE_CHECKING, Any

from pydantic import Field

from semantic_kernel.processes.kernel_process.kernel_process_edge import KernelProcessEdge
from semantic_kernel.processes.kernel_process.kernel_process_state import KernelProcessState
from semantic_kernel.processes.kernel_process.kernel_process_step_info import KernelProcessStepInfo
from semantic_kernel.processes.process_state_metadata_utils import kernel_process_to_process_state_metadata
from semantic_kernel.utils.feature_stage_decorator import experimental

if TYPE_CHECKING:
    from semantic_kernel.processes.kernel_process.kernel_process_edge import KernelProcessEdge
    from semantic_kernel.processes.kernel_process.kernel_process_step_state_metadata import KernelProcessStateMetadata


@experimental
class KernelProcess(KernelProcessStepInfo):
    """A kernel process."""

    steps: list[KernelProcessStepInfo] = Field(default_factory=list)
    factories: dict[str, Callable] = Field(default_factory=dict)

    def __init__(
        self,
        state: KernelProcessState,
        steps: list[KernelProcessStepInfo],
        edges: dict[str, list["KernelProcessEdge"]] | None = None,
        factories: dict[str, Callable] | None = None,
    ):
        """Initialize the kernel process.

        Args:
            state: The state of the process.
            steps: The steps of the process.
            edges: The edges of the process. Defaults to None.
            factories: The factories of the process. This allows for the creation of
                steps that require complex dependencies that cannot be JSON serialized or deserialized.
        """
        if not state:
            raise ValueError("state cannot be None")
        if not steps:
            raise ValueError("steps cannot be None")
        if not state.name:
            raise ValueError("state.Name cannot be None")

        process_steps = []
        process_steps.extend(steps)

        args: dict[str, Any] = {
            "steps": process_steps,
            "inner_step_type": KernelProcess,
            "state": state,
            "output_edges": edges or {},
        }

        if factories:
            args["factories"] = factories

        super().__init__(**args)

    def to_process_state_metadata(self) -> "KernelProcessStateMetadata":
        """Converts a kernel process to process state metadata."""
        return kernel_process_to_process_state_metadata(self)

File: nodejs/semantic-kernel/processes/kernel-process/kernel-process.ts
import { kernelProcessToProcessStateMetadata } from '../process-state-metadata-utils'
import type { KernelProcessEdge } from './kernel-process-edge'
import { KernelProcessState } from './kernel-process-state'
import { KernelProcessStepInfo } from './kernel-process-step-info'
import type { KernelProcessStateMetadata } from './kernel-process-step-state-metadata'

/**
 * A kernel process.
 * @experimental
 */
export class KernelProcess extends KernelProcessStepInfo {
  /**
   * The steps in the process.
   */
  steps: KernelProcessStepInfo[] = []

  /**
   * The factories for creating steps with complex dependencies.
   */
  factories: Record<string, (...args: any[]) => any> = {}

  constructor(params: {
    state: KernelProcessState
    steps: KernelProcessStepInfo[]
    edges?: Record<string, KernelProcessEdge[]> | null
    factories?: Record<string, (...args: any[]) => any> | null
  }) {
    if (!params.state) {
      throw new Error('state cannot be null')
    }
    if (!params.steps || params.steps.length === 0) {
      throw new Error('steps cannot be null')
    }
    if (!params.state.name) {
      throw new Error('state.name cannot be null')
    }

    const processSteps = [...params.steps]

    super({
      innerStepType: KernelProcess,
      state: params.state,
      outputEdges: params.edges || {},
    })

    this.steps = processSteps

    if (params.factories) {
      this.factories = params.factories
    }
  }

  /**
   * Converts a kernel process to process state metadata.
   * @returns The process state metadata
   */
  toProcessStateMetadata(): KernelProcessStateMetadata {
    return kernelProcessToProcessStateMetadata(this)
  }
}
